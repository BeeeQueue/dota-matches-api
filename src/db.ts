import { InsertObject, Kysely } from "kysely"
import { D1Dialect } from "kysely-d1"
import { pick } from "remeda"

import type { Match, Team } from "./dota"

export type MatchTable = {
  id: string
  matchType: string | null
  teamOneId: string | null
  teamTwoId: string | null
  leagueName: string | null
  streamUrl: string | null
  startsAt: string | null
}

export type LeagueTable = {
  name: string
  url: string | null
}

export type TeamTable = {
  id: string
  name: string
  url: string | null
}

type Database = {
  match: MatchTable
  league: LeagueTable
  team: TeamTable
}

export type Db = Kysely<Database>

export const createDb = (env: Env) =>
  new Kysely<Database>({
    dialect: new D1Dialect({ database: env.__D1_BETA__MATCHES }),
  })

export const getTeamsFromDb = async (db: Db): Promise<Team[]> => {
  return await db.selectFrom("team").select(["name", "url"]).execute()
}

export const upsertTeamsData = async (db: Db, teams: Team[]) => {
  return await Promise.all(
    [...teams.values()].map(({ name, url }) =>
      db
        .insertInto("team")
        .values({ id: name!, name: name!, url })
        .onConflict((b) => b.column("id").doUpdateSet({ id: name!, name: name!, url }))
        .execute(),
    ),
  )
}

export const getMatchDataFromDb = async (db: Db): Promise<Match[]> => {
  const matchData = await db
    .selectFrom("match")
    .leftJoin("league", "match.leagueName", "league.name")
    .leftJoin("team as teamOne", "match.teamOneId", "teamOne.id")
    .leftJoin("team as teamTwo", "match.teamTwoId", "teamTwo.id")
    .select([
      "match.id",
      "matchType",
      "streamUrl",
      "startsAt",
      "leagueName",
      "league.url as leagueUrl",
      "teamOneId",
      "teamOne.name as teamOneName",
      "teamOne.url as teamOneUrl",
      "teamTwoId",
      "teamTwo.name as teamTwoName",
      "teamTwo.url as teamTwoUrl",
    ])
    .execute()

  return matchData.map((match) => ({
    hash: match.id!,
    ...pick(match, ["matchType", "streamUrl", "startsAt", "leagueName", "leagueUrl"]),
    teams: [
      { name: match.teamOneName, url: match.teamOneUrl },
      { name: match.teamTwoName, url: match.teamTwoUrl },
    ],
  }))
}

export const upsertMatchData = async (db: Db, matches: Match[]) => {
  console.log("Upserting match data...")

  console.log("Deleting old matches...")
  const deletePromise = db.deleteFrom("match").execute()

  console.log("Generating new data...")
  const matchData = matches.map<InsertObject<Database, "match">>((match) => ({
    id: match.hash,
    matchType: match.matchType,
    teamOneId: match.teams[0]?.name ?? null,
    teamTwoId: match.teams[1]?.name ?? null,
    leagueName: match.leagueName,
    streamUrl: match.streamUrl,
    startsAt: match.startsAt,
  }))

  const teams = new Map<string, InsertObject<Database, "team">>()
  for (const match of matches) {
    if (match.teams[0]?.name != null && !teams.has(match.teams[0].name)) {
      teams.set(match.teams[0].name, {
        id: match.teams[0].name,
        name: match.teams[0].name,
        url: match.teams[0].url,
      })
    }
    if (match.teams[1]?.name != null && !teams.has(match.teams[1].name)) {
      teams.set(match.teams[1].name, {
        id: match.teams[1].name,
        name: match.teams[1].name,
        url: match.teams[1].url,
      })
    }
  }

  const leagues = new Map<string, InsertObject<Database, "league">>()
  for (const match of matches) {
    if (match.leagueName == null || leagues.has(match.leagueName)) continue

    leagues.set(match.leagueName, {
      name: match.leagueName,
      url: match.leagueUrl,
    })
  }

  console.log("Waiting for deletion to finish...")
  await deletePromise

  console.log("Inserting matches...")
  await db.insertInto("match").values(matchData).execute()

  console.log("Inserting teams...")
  await Promise.all(
    [...teams.values()].map((team) =>
      db
        .insertInto("team")
        .values(team)
        .onConflict((b) => b.column("id").doUpdateSet(team))
        .execute(),
    ),
  )

  console.log("Inserting leagues...")
  await Promise.all(
    [...leagues.values()].map((league) =>
      db
        .insertInto("league")
        .values(league)
        .onConflict((b) => b.column("name").doUpdateSet(league))
        .execute(),
    ),
  )
}
