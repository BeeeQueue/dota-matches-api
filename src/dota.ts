import { mande } from "mande"
import { HTMLElement, parse } from "node-html-parser"
import PQueue from "p-queue"

const THREE_HOURS = 60 * 60 * 3

type Team = {
  name: string | null
  url: string | null
}

export type Match = {
  teams: [Team, Team]
  matchType: string | null
  startsAt: Date | null
  streamUrl: string | null
}

const liquipediaQueue = new PQueue({ intervalCap: 1, interval: 30_000 })
const liquipediaClient = mande("https://liquipedia.net/dota2", {
  responseAs: "json",
  query: {
    format: "json",
  },
  headers: {
    /* eslint-disable @typescript-eslint/naming-convention */
    "Accept-Encoding": "gzip",
  },
})

type LiquipediaBody = {
  parse: {
    title: string
    pageid: number
    revid: number
    text: {
      ["*"]: string
    }
    langlinks: string[]
    categories: unknown[]
    links: unknown[]
    templates: unknown[]
    images: string[]
    externallinks: string[]
  }
}

const extractTeam = (team$: HTMLElement): Team => {
  const name = team$
    .querySelector(".team-template-text > a")
    ?.attrs?.title?.replace(/\(.*?\)/g, "")
    ?.trim()
  const urlPath = team$.querySelector("[href^='/dota2/']")?.attrs?.href

  return {
    name: name ?? null,
    url: urlPath ? `https://liquipedia.com${urlPath}` : null,
  }
}

const fetchTeamsData = async (country: string) => {
  const data = await liquipediaQueue.add(() =>
    liquipediaClient.get<LiquipediaBody>("/api.php", {
      headers: {
        "User-Agent": `dota-matches-api-${country}/${GIT_SHA}`,
      },
      query: {
        action: "parse",
        page: "Liquipedia:Upcoming_and_ongoing_matches",
      },
    }),
  )

  const root = parse(data.parse.text["*"])

  const $matches = root.querySelectorAll("[data-toggle-area-content='2'] > table")
  if ($matches.length === 0) return []

  const matches = $matches.map<Match>(($match) => {
    const teamLeft$ = $match.querySelector(".team-left")!
    const teamRight$ = $match.querySelector(".team-right")!
    const versus$ = $match.querySelector(".versus")!
    const meta$ = $match.querySelector(".timer-object")!

    const matchType = versus$.querySelector("abbr")?.textContent
    const startTime = meta$.attrs["data-timestamp"]
    const streamName = meta$.attrs["data-stream-twitch"]

    return {
      teams: [extractTeam(teamLeft$), extractTeam(teamRight$)],
      matchType: matchType ?? null,
      startsAt: startTime ? new Date(Number(startTime) * 1000) : null,
      streamUrl: streamName ? `https://www.twitch.tv/${streamName}` : null,
    }
  })

  return matches
}

const CACHE_KEY = "liquipedia-matches"

const getMatches = async (env: Env, country: string): Promise<Match[]> => {
  console.log(`Getting match data...`)

  const cached = await env.CACHE.get(CACHE_KEY)
  if (cached != null) {
    return JSON.parse(cached)
  }

  const matches = await fetchTeamsData(country)

  await env.CACHE.put(CACHE_KEY, JSON.stringify(matches), {
    expirationTtl: THREE_HOURS,
  })

  return matches
}

export const Dota = { getMatches }
