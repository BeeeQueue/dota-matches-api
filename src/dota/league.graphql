query League($id: Int!) {
  league(id: $id) {
    displayName
    region
    stats {
      matchCount
    }
    nodeGroups {
      nodes {
        scheduledTime
        hasStarted
        isCompleted
        nodeType

        teamOneWins
        teamTwoWins
        teamOne {
          id
          name
        }
        teamTwo {
          id
          name
        }

        matches {
          id
        }
      }
    }
  }
}
