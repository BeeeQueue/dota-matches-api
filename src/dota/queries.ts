export const teamsQuery = /* GraphQL */ `
  query Teams($ids: [Int!]!) {
    teams(teamIds: $ids) {
      id
      name
      series(request: { take: 20, skip: 0 }) {
        id
        type
        matches {
          startDateTime
        }

        teamOne {
          id
          name
        }
        teamTwo {
          id
          name
        }

        league {
          displayName
        }
      }
    }
  }
`
