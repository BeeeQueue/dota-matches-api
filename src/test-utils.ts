import type { MockAgent } from "undici"

import type { Db, SubscriptionTable, TeamTable } from "./db"

declare module "vitest" {
  // eslint-disable-next-line @typescript-eslint/consistent-type-definitions
  interface TestContext {
    agent: MockAgent
    env: Env
    db: Db
  }
}

export const GUILD_ID = "987613986523"
export const CHANNEL_ID = "0986526095326812"

export const createSub = (teamName: string): SubscriptionTable => ({
  guildId: GUILD_ID,
  channel: CHANNEL_ID,
  teamName,
})

export const createTeam = (teamName: string): TeamTable => ({
  id: teamName,
  name: teamName,
  url: "url",
})
