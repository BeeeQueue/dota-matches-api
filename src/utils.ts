import ms, { StringValue } from "ms"

import { RequestInitExStatus, ok } from "@worker-tools/response-creators"

export const nowSeconds = () => Math.round(Date.now() / 1000)

export const seconds = (input: StringValue) => Math.round(ms(input) / 1000)

export const json = <T = unknown>(data: T, init: RequestInitExStatus = {}) =>
  ok(JSON.stringify(data, null, 2), {
    headers: {
      ...init.headers,
      "Content-Type": "application/json",
    },
    ...init,
  })
