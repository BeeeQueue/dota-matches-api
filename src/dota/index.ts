import { differenceInHours } from "date-fns"
import { mande, MandeError } from "mande"

import { isNotNil } from "../utils"

import { teamsQuery } from "./queries"
import { TeamsQuery, TeamsQueryVariables } from "./types"

export type Match = {
  id: number
  startsAt: Date
  teams: [
    {
      id: number
      name: string
    },
    {
      id: number
      name: string
    },
  ]
}

const isMatchRelevant = (match: Match): boolean =>
  Math.abs(differenceInHours(new Date(), match.startsAt)) <= 24

const cleanMatchData = (
  series: NonNullable<
    NonNullable<NonNullable<TeamsQuery["teams"]>[number]>["series"]
  >[number],
): Match | null => {
  if (series == null) return null

  const firstGameStartTime = series.matches?.reduce(
    (accum, match) =>
      (accum ?? Number.NEGATIVE_INFINITY) <
      (match!.startDateTime ?? Number.POSITIVE_INFINITY)
        ? match!.startDateTime
        : accum,
    null as number | null,
  )

  if (firstGameStartTime == null) return null

  return {
    id: series.id,
    startsAt: new Date(firstGameStartTime * 1000),
    teams: [
      {
        id: series.teamOne!.id,
        name: series.teamOne!.name!,
      },
      {
        id: series.teamTwo!.id,
        name: series.teamTwo!.name!,
      },
    ],
  }
}

const stratzClient = mande("https://api.stratz.com", {
  responseAs: "json",
  headers: {
    Authorization: `Bearer ${STRATZ_TOKEN}`,
  },
})

const fetchLeagueMatches = async (ids: number[]): Promise<TeamsQuery> => {
  if (ids.length === 0) {
    return {
      teams: [],
    }
  }

  console.log(`Fetching ${ids.join(", ")}...`)

  const variables: TeamsQueryVariables = {
    ids,
  }
  const result = await stratzClient
    .post<{ data: TeamsQuery }>("/graphql", {
      query: teamsQuery,
      variables,
    })
    .catch((error: MandeError) => error)

  if (result instanceof Error) {
    const body = await result.response
      .json<Record<string, unknown>>()
      .catch(() => result.response.text().catch(() => null))
    console.error({
      status: result.response.status,
      body,
    })
    throw new Error(
      `Failed to fetch matches: ${result.response.status}, ${JSON.stringify(body!)}`,
    )
  }

  return result.data
}

const getTeamsData = async (env: Env, ids: number[]) => {
  console.log(`Getting match data for ${ids.join(", ")}...`)

  const matches: Match[] = []
  const notCachedIds = new Set<number>()

  await Promise.all(
    ids.map(async (id) => {
      const cached = await env.CACHE.get(id.toString())
      if (cached != null) {
        matches.push(...(JSON.parse(cached) as Match[]))
        return
      }

      notCachedIds.add(id)
    }),
  )
  console.log(`Found ${ids.length - notCachedIds.size} cached teams`)

  if (notCachedIds.size > 0) {
    console.log(`Fetching ${notCachedIds.size} matches...`)

    const data = await fetchLeagueMatches([...notCachedIds])
    const matchesByTeam = (data.teams ?? []).reduce((accum, team) => {
      if (team == null || team.series == null) return accum

      accum[team.id.toString()] ??= []

      const newMatches = team.series
        .map(cleanMatchData)
        .filter((match): match is Match => isNotNil(match) && isMatchRelevant(match))
      accum[team.id.toString()].push(...newMatches)

      return accum
    }, {} as Record<string, Match[]>)

    console.log(`Caching ${Object.keys(matchesByTeam).length} teams...`)
    await Promise.all(
      Object.entries(matchesByTeam).map(async ([teamId, newMatches]) => {
        await env.CACHE.put(teamId, JSON.stringify(newMatches), {
          expirationTtl: 60 * 60 * 3,
        })

        matches.push(...newMatches)
      }),
    )
  }

  console.log(`Deduping and returning...`)
  const matchIds = new Set<number>()
  return matches.filter((match) => {
    if (!matchIds.has(match.id)) {
      matchIds.add(match.id)
      return true
    }

    return false
  })
}

export const Dota = { getTeamsData }
