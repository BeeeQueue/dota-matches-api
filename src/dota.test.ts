import { setTimeout } from "timers/promises"

import ms from "ms"
import { beforeEach, expect, it, vi } from "vitest"

import { createDb, Db } from "./db"
import { getTeams, LiquipediaBody, parseTeamsPage } from "./dota"
import teamsFixture from "./fixtures/teams.html?raw"
import { initDb } from "./test-utils"
import { MetaKey } from "./utils"

const describe = setupMiniflareIsolatedStorage()
const agent = getMiniflareFetchMock()

describe("parseTeamsPage", () => {
  it("correctly parses the body", () => {
    const result = parseTeamsPage(teamsFixture)

    expect(result).toBeDefined()
    expect(result).toMatchSnapshot()
  })
})

describe("getTeams", () => {
  let env: Env
  let db: Db

  beforeEach(async () => {
    vi.resetAllMocks()

    vi.setSystemTime(new Date("2020-01-01"))

    agent.disableNetConnect()
    agent.assertNoPendingInterceptors()

    env = getMiniflareBindings()
    db = createDb(env)
    await initDb(env)
  })

  const body: LiquipediaBody = {
    parse: {
      text: {
        "*": teamsFixture,
      },
    } as Partial<LiquipediaBody["parse"]> as LiquipediaBody["parse"],
  }
  const mockApiRequest = () => {
    const handler = vi.fn(() => body)

    agent
      .get("https://liquipedia.net")
      .intercept({
        path: /dota2.api\.php/,
        method: "GET",
      })
      .reply(200, handler)

    return handler
  }

  it("fetches teams from api if not cached", async () => {
    const apiHandler = mockApiRequest()

    const result = await getTeams(env, db)("test")

    // New values
    expect(result).toContain("OG")
    expect(result).toContain("Team Liquid")

    expect(apiHandler).toHaveBeenCalledOnce()

    const data = await db.selectFrom("team").selectAll().execute()
    expect(data[0]).toStrictEqual({
      id: "5ManMidas",
      name: "5ManMidas",
      url: "https://liquipedia.net/dota2/5ManMidas",
    })
  })

  it("fetches teams from cache if cached", async () => {
    await db.insertInto("team").values({ id: "OG", name: "OG", url: "url" }).execute()
    await env.META.put(MetaKey.TEAMS_LAST_FETCHED, (Date.now() - ms("10m")).toString())

    const result = await getTeams(env, db)("test")

    // Cached values
    expect(result).toContain("OG")
    expect(result).not.toContain("Team Liquid")
  })

  it("fetches teams from cache and updates it if soft expired", async () => {
    const apiHandler = mockApiRequest()

    await db.insertInto("team").values({ id: "OG", name: "OG", url: "url" }).execute()
    const l = (Date.now() - ms("6d")).toString()
    await env.META.put(MetaKey.TEAMS_LAST_FETCHED, l)

    const result = await getTeams(env, db)("test")

    // Cached values
    expect(result).toContain("OG")
    expect(result).not.toContain("Team Liquid")

    await setTimeout(100)
    // Still fetched new ones
    expect(apiHandler).toHaveBeenCalledOnce()
    const data = await db.selectFrom("team").selectAll().orderBy("name").execute()
    expect(data[0]).toStrictEqual({
      id: "5ManMidas",
      name: "5ManMidas",
      url: "https://liquipedia.net/dota2/5ManMidas",
    })
  })
})
