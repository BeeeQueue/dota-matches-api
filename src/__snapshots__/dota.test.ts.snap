// Vitest Snapshot v1

exports[`parseTeamsPage > correctly parses the body 1`] = `
[
  "5ManMidas",
  "5RATFORCESTAFF",
  "Alliance",
  "Arkosh Gaming",
  "Army Geniuses",
  "Aster.Aries",
  "Balrogs",
  "beastcoast",
  "BetBoom Team",
  "BOOM Esports",
  "Brame",
  "CDEC Gaming",
  "cybercats",
  "Dandelion Esport Club",
  "Dreamers Esports",
  "EHOME",
  "Entity",
  "Evil Geniuses",
  "Execration",
  "felt",
  "Fnatic",
  "Gaimin Gladiators",
  "goonsquad",
  "Gorillaz-Pride",
  "HellRaisers",
  "Hokori",
  "HYDRA",
  "iG Vitality",
  "Infamous",
  "Infinity",
  "Into The Breach",
  "Invictus Gaming",
  "IVY",
  "jfshfh178",
  "KBU.US",
  "Lava",
  "LBZS",
  "Lilgun",
  "Mad Kings",
  "Mind Games",
  "Natus Vincere",
  "Nemiga Gaming",
  "Neon Esports",
  "Nigma Galaxy",
  "Nigma Galaxy SEA",
  "NoPing e-sports",
  "nouns",
  "OG",
  "Omega Gaming",
  "One Move",
  "Outsiders",
  "Pari Parni",
  "Pavaga Gaming",
  "Polaris Esports",
  "PSG.LGD",
  "PuckChamp",
  "Royal Never Give Up",
  "RSG",
  "SG esports",
  "ShenZhen",
  "simply TOOBASED",
  "Soniqs",
  "Sparking Arrow Gaming",
  "T1",
  "Talon Esports",
  "Team Aster",
  "Team Bald Reborn",
  "Team DogChamp",
  "Team Liquid",
  "Team MagMa",
  "Team Secret",
  "Team SMG",
  "Team Unique",
  "Tempest",
  "The Cut",
  "Thunder Awaken",
  "TNC Predator",
  "TSM",
  "Tundra Esports",
  "Vici Gaming",
  "Vicious Gaming",
  "Virtus.pro",
  "Wildcard Gaming",
  "Wolf Team",
  "Xtreme Gaming",
  "Yangon Galacticos",
  "Ybb Gaming",
]
`;
